File: python/tokenize.py
from stream import *

class Token():
    def _membersStr(self):
        return ", ".join((f'{k}={repr(v)}' for k, v in self.__dict__.items()))
    def __repr__(self):
        return f'{type(self).__name__}({self._membersStr()})'

class Integer(Token):
    def __init__(self, value):
        self.value = value
class Float(Token):
    def __init__(self, value):
        self.value = value
class Identifier(Token):
    def __init__(self, value):
        self.value = value
class Plus(Token):
    pass
class Minus(Token):
    pass
class Times(Token):
    pass
class Divide(Token):
    pass
class Mod(Token):
    pass
class Pipe(Token):
    pass
class LParen(Token):
    pass
class RParen(Token):
    pass
class LBrace(Token):
    pass
class RBrace(Token):
    pass
class LBracket(Token):
    pass
class RBracket(Token):
    pass
class Comma(Token):
    pass
class Dot(Token):
    pass
class Elipsis          (Token):
    pass
class Epsilon          (Token):
    pass
class Indent           (Token):
    pass
class UnIndent         (Token):
    pass
class EOL              (Token):
    pass
class Equals           (Token):
    pass
class Arrow            (Token):
    pass
class SingleUnderscore (Token):
    pass
class Is               (Token):
    pass
class Not              (Token):
    pass
class In               (Token):
    pass
class Update           (Token):
    pass
class For              (Token):
    pass
class To               (Token):
    pass
class And              (Token):
    pass
class Or               (Token):
    pass
class NotEquals        (Token):
    pass
class Question         (Token):
    pass
class Maybe            (Token):
    pass
class Struct           (Token):
    pass
class Union            (Token):
    pass
class Type             (Token):
    pass
class Interface        (Token):
    pass
class Set              (Token):
    pass
class Module           (Token):
    pass
class Def              (Token):
    pass
class Const            (Token):
    pass
class Assign           (Token):
    pass
class If               (Token):
    pass
class Match            (Token):
    pass
class Case             (Token):
    pass
class Return           (Token):
    pass
class Pound            (Token):
    pass

def isWhitespace(el):
    return el == '\n' or el == ' '

def span(string, func):
    for i, el in enumerate(string):
        if not func(el):
            return string[:i], string[i:]
    return string, '' 
def findRightmost(string: str, el):
    i = string.rfind(el)
    return i if i >= 0 else None
def find(string: str, el):
    i = string.find(el)
    return i if i >= 0 else None

def isDigit(string):
    return ord('0') <= ord(string) <= ord('9')

def isAlpha(string):
    return ord('a') <= ord(string) <= ord('z') or ord('A') <= ord(string) <= ord('Z')
def isIdentifierBegining(string):
    return isAlpha(string) or string == '_'
def isIdentifier(string):
    return isDigit(string) or isAlpha(string) or string == '_'

def tokenizeIdentifier(string):
    match string:
        case 'is':
            return Is()
        case 'not':
            return Not()
        case 'in':
            return In()
        case 'update':
            return Update()
        case 'for':
            return For()
        case 'to':
            return To()
        case 'and':
            return And()
        case 'or':
            return Or()
        case 'struct':
            return Struct()
        case 'union':
            return Union()
        case 'type':
            return Type()
        case 'interface':
            return Interface()
        case 'set':
            return Set()
        case 'module':
            return Module()
        case 'def':
            return Def()
        case 'const':
            return Const()
        case 'if':
            return If()
        case 'match':
            return Match()
        case 'case':
            return Case()
        case 'maybe':
            return Maybe()
        case 'return':
            return Return()
        case default:
            return Identifier(default)

@streamFunc
def Tokenize(input: str, tabs: int) -> Stream:
    if input == '':
        return empty
    match input[0]:
        case '+':
            return Plus() * Tokenize(input[1:], tabs)
        case '-':
            return Minus() * Tokenize(input[1:], tabs)
        case '*':
            return Times() * Tokenize(input[1:], tabs)
        case '/':
            return Divide() * Tokenize(input[1:], tabs)
        case '%':
            return Mod() * Tokenize(input[1:], tabs)
        case '|':
            return Pipe() * Tokenize(input[1:], tabs)
        case '(':
            return LParen() * Tokenize(input[1:], tabs)
        case ')':
            return RParen() * Tokenize(input[1:], tabs)
        case '{':
            return LBrace() * Tokenize(input[1:], tabs)
        case '}':
            return RBrace() * Tokenize(input[1:], tabs)
        case '[':
            return LBracket() * Tokenize(input[1:], tabs)
        case ']':
            return RBracket() * Tokenize(input[1:], tabs)
        case ',':
            return Comma() * Tokenize(input[1:], tabs)
        case '.':
            if input.startswith('...'):
                return Elipsis() * Tokenize(input[3:], tabs)
            return Dot() * Tokenize(input[1:], tabs)
        case el if isWhitespace(el):
            wspace, rest = span(input, isWhitespace)
            index = findRightmost(input, '\n')
            if index is None:
                return Tokenize(input[len(wspace):], tabs)
            spaces = wspace[index+1:]
            numSpaces = len(spaces)
            if numSpaces % 4 != 0:
                assert False and 'indentation must be multiple of 4'
            numTabs = numSpaces // 4
            if numTabs > tabs:
                return chain(
                    [EOL()], 
                    (Indent() for _ in range(numTabs - tabs)), 
                    Tokenize(rest, numSpaces),
                )
            if numTabs < tabs:
                return chain(
                    [EOL()],
                    (UnIndent() for _ in range(tabs - numTabs)),
                    Tokenize(rest, numSpaces)
                )
            return EOL() * Tokenize(rest, numSpaces)
        case '=':
            if input.startswith('=='):
                return Equals() * Tokenize(input[2:], tabs)
            if input.startswith('=>'):
                return Arrow() * Tokenize(input[2:], tabs)
            return Assign() * Tokenize(input[1:], tabs)
        case '!':
            if not input.startswith('!='):
                assert False and 'unexpected token !, not followed by `=`'
            return NotEquals() * Tokenize(input[2:], tabs)
        case '?':
            if input.startswith('??'):
                return Maybe() * Tokenize(input[2:], tabs)
            return Question() * Tokenize(input[1:], tabs)
        case '#':
            index = find(input, '\n')
            if index is None:
                return empty 
            return Tokenize(input[index+1:], tabs)
        case el if isDigit(el):
            digits, rest = span(input, isDigit)
            if not rest.startswith('.'):
                return Integer(digits) * Tokenize(rest, tabs)
            moreDigits, rest = span(rest[1:], isDigit)
            return Float(digits + '.' + moreDigits) * Tokenize(rest, tabs)
        case el if isIdentifierBegining(el):
            ident, rest = span(input, isIdentifier)
            return tokenizeIdentifier(ident) * Tokenize(rest, tabs)

print(*Tokenize('if i == 23 (wow) => cool + 27', 0))

File: haskell/tokenize.hs
{-# LANGUAGE OverloadedStrings #-}

module NasperTokenizer where

import Data.Char (isDigit, isAlpha, isSpace)
import Data.Text (Text)
import qualified Data.Text as T

data Token
  = Integer String
  | Float String
  | Identifier String
  | Plus | Minus | Times | Divide | Mod
  | Pipe | LParen | RParen
  | LBrace | RBrace | LBracket | RBracket
  | Comma | Dot | Ellipsis | Epsilon
  | Indent | UnIndent | Equals | Arrow | Assign
  | SingleUnderscore | Is | Not | In
  | Update | For | To | And | Or
  | Question | MaybeToken | Struct | Union | Type
  | Interface | Set | Module | Def | Const
  | If | Match | Case | Pound | Return
  deriving (Show, Eq)

tokenize :: Text -> Int -> [Token]
tokenize "" n = []
tokenize input tabs
  | T.isPrefixOf "+" input = Plus : tokenize (T.tail input) tabs
  | T.isPrefixOf "-" input = Minus : tokenize (T.tail input) tabs
  | T.isPrefixOf "*" input = Times : tokenize (T.tail input) tabs
  | T.isPrefixOf "/" input = Divide : tokenize (T.tail input) tabs
  | T.isPrefixOf "%" input = Mod : tokenize (T.tail input) tabs
  | T.isPrefixOf "|" input = Pipe : tokenize (T.tail input) tabs
  | T.isPrefixOf "(" input = LParen : tokenize (T.tail input) tabs
  | T.isPrefixOf ")" input = RParen : tokenize (T.tail input) tabs
  | T.isPrefixOf "{" input = LBrace : tokenize (T.tail input) tabs
  | T.isPrefixOf "}" input = RBrace : tokenize (T.tail input) tabs
  | T.isPrefixOf "[" input = LBracket : tokenize (T.tail input) tabs
  | T.isPrefixOf "]" input = RBracket : tokenize (T.tail input) tabs
  | T.isPrefixOf "," input = Comma : tokenize (T.tail input) tabs
  | T.isPrefixOf "..." input = Ellipsis : tokenize (T.drop 3 input) tabs
  | T.isPrefixOf "." input = Dot : tokenize (T.tail input) tabs
  | T.isPrefixOf "--" input = tokenize (T.dropWhile (/= '\n') input) tabs
  | T.isPrefixOf "==" input = Equals : tokenize (T.drop 2 input) tabs
  | T.isPrefixOf "=>" input = Arrow : tokenize (T.drop 2 input) tabs
  | T.isPrefixOf "=" input = Assign : tokenize (T.tail input) tabs
  | isDigit (T.head input) =
      let (num, rest) = T.span isDigit input
      in if T.isPrefixOf "." rest
         then let (frac, rest') = T.span isDigit (T.tail rest)
              in Float (T.unpack num <> "." <> T.unpack frac) : tokenize rest' tabs
         else Integer (T.unpack num) : tokenize rest tabs
  | isAlpha (T.head input) || T.head input == '_' =
      let (ident, rest) = T.span (\c -> isAlpha c || isDigit c || c == '_') input
      in case ident of
          "is" -> Is : tokenize rest tabs
          "not" -> Not : tokenize rest tabs
          "in" -> In : tokenize rest tabs
          "update" -> Update : tokenize rest tabs
          "for" -> For : tokenize rest tabs
          "to" -> To : tokenize rest tabs
          "and" -> And : tokenize rest tabs
          "or" -> Or : tokenize rest tabs
          "struct" -> Struct : tokenize rest tabs
          "union" -> Union : tokenize rest tabs
          "type" -> Type : tokenize rest tabs
          "interface" -> Interface : tokenize rest tabs
          "set" -> Set : tokenize rest tabs
          "module" -> Module : tokenize rest tabs
          "def" -> Def : tokenize rest tabs
          "const" -> Const : tokenize rest tabs
          "if" -> If : tokenize rest tabs
          "match" -> Match : tokenize rest tabs
          "case" -> Case : tokenize rest tabs
          "maybe" -> MaybeToken : tokenize rest tabs
          "return" -> Return : tokenize rest tabs
          _ -> Identifier (T.unpack ident) : tokenize rest tabs
  | isSpace (T.head input) =
      let spaces = T.takeWhile isSpace input
          numSpaces = T.length spaces
          numTabs = numSpaces `div` 4
          rest = T.dropWhile isSpace input
      in if numTabs > tabs
          then replicate (numTabs - tabs) Indent ++ tokenize rest numTabs
          else if numTabs < tabs
              then replicate (tabs - numTabs) UnIndent ++ tokenize rest numTabs
              else tokenize rest tabs
  | otherwise = tokenize (T.tail input) tabs -- Skip unknown character

-- Example usage
main :: IO ()
main = print $ tokenize "\
\if x == 4 \
\   return y \
\else\
\    print(1,2, 3)\
\    return 5\
\" 0
